### PR TITLE
Parameterize SmartTransactionsController state by ChainId for MultiChain + Integrate PollingController Mixin

### PR DESCRIPTION
**Pull request by adonesky1**
~TODO:~
- [x] ~E2E test with extension to ensure working properly before taking out of draft~

Working e2e test here:
https://github.com/MetaMask/smart-transactions-controller/assets/34557516/307de2da-1fe3-43c8-8faa-7bf51da28eaa

Resolves: https://github.com/MetaMask/MetaMask-planning/issues/1039

Integrates new `PollingController` abstraction into the `SmartTransactionsController`, implements the `_executePoll` method and parameterizes the `updateSmartTransactions` method by `networkClientId` such that pending SmartTransaction statuses can be polled for concurrently across different chains.

A note: As is our current controller refactor strategy, these changes are additive, backwards compatible and coexist alongside the existing globally selected network pattern. Once we fully adopt this polling pattern we should no longer access the root liveness and fees state but rather access these values by chainId from `livenessByChainId` and `feesByChainId`.

### Added
- Integrates `PollingController` mixin into the `SmartTransactionsController` and implements the abstract `_executePoll` method.
### Changed
- Adds optional options object containing a `networkClientId` argument to the `updateSmartTransaction` method which, if passed, is used fetch the chainId that corresponds to the networkClientId and is then used to fetch and update the status of pending smart transactions for that chainId.

### CODE REVIEW

File: src/SmartTransactionsController.ts
Comment (MajorLift) [inline src/SmartTransactionsController.ts:527]:
```suggestion
    const data = (await this.fetch(url)) as Record<
      string,
      SmartTransactionsStatus
    >;
```

File: src/SmartTransactionsController.ts
Comment (MajorLift) [inline src/SmartTransactionsController.ts:535]:
```suggestion
          statusMetadata: stxStatus,
          status: calculateStatus(stxStatus),
          cancellable: isSmartTransactionCancellable(stxStatus),
```

Comment (MajorLift) [reply src/SmartTransactionsController.ts:535]: Follows from https://github.com/MetaMask/smart-transactions-controller/pull/237/files#r1402633206, which lets `stxStatus` be inferred.

File: src/SmartTransactionsController.ts
Comment (MajorLift) [inline src/SmartTransactionsController.ts:516]: This return type seems to be inaccurate.

```suggestion
  ): Promise<Record<string, SmartTransactionsStatus>> {
```

File: src/SmartTransactionsController.ts
Comment (MajorLift) [inline src/SmartTransactionsController.ts:408]:
```suggestion
    const smartTransactionsForChainId = smartTransactions[chainId];
```

This can be removed without causing errors.

File: src/SmartTransactionsController.ts
Comment (MajorLift) [inline src/SmartTransactionsController.ts:109]:
```suggestion
      provider: Provider;
```

This can be done without disrupting anything, since the errors caused by this are already suppressed by `// @ts-expect-error TODO: Provider type alignment` directives.

File: src/test-helpers.ts
Comment (MajorLift) [inline src/test-helpers.ts:25]: Nit: just in case `duration` is passed into something that expects an integer.
```suggestion
  stepSize = Math.round(duration / 4),
```

File: package.json
Comment (MajorLift) [inline package.json:31]:
```suggestion
    "@metamask/base-controller": "^4.0.0",
    "@metamask/controller-utils": "^6.1.0",
    "@metamask/eth-query": "^4.0.0",
```

With these version bumps (and https://github.com/MetaMask/smart-transactions-controller/pull/237#discussion_r1402649548), all `// @ts-expect error TODO: Provider type alignment` directives can be removed.

File: src/test-helpers.ts
Comment (adonesky1) [reply src/test-helpers.ts:25]: It will only ever be passed to `clock.tickAsync` but I suppose this is fine

Comment (adonesky1) [reply src/test-helpers.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/93a05f14cbf91a344332dcf07d39f6924d5eacd3

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/93a05f14cbf91a344332dcf07d39f6924d5eacd3

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/93a05f14cbf91a344332dcf07d39f6924d5eacd3

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/93a05f14cbf91a344332dcf07d39f6924d5eacd3

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/93a05f14cbf91a344332dcf07d39f6924d5eacd3

File: src/SmartTransactionsController.test.ts
Comment (jiexi) [inline src/SmartTransactionsController.test.ts:17]:
```suggestion
```

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:297]:
```suggestion
    let { ethQuery, config: { chainId } = this;
```

ehh idk. `let { ethQuery } = this;` specifically feels funky, but i guess that's only because you aren't usually deconstructing `this`

Comment (jiexi) [reply src/SmartTransactionsController.ts:297]: maybe even `let chainId, ethQuery;` since you default it in `#updateSmartTransaction` already

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:353]: `nextSmartTransactions` would be undefined in state rather than an empty array if  `currentSmartTransactions` is undefined. Is that what we want?

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:302]: are these helpers useful here?

```
    const chainId = this.getChainId({ networkClientId });
    const ethQuery = this.getEthQuery({ networkClientId });
```

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:616]: you only want to set `fees` if it matches the networkClientId matches the current globally selected chainId, right?

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:708]: does this need a default value `= {}`

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:751]: `liveness` should only be set if the networkClientId's chainId matches the globally selected chainId

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts:121]: 🤔 yeah don't know why I did this. Will revert.

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts:297]: Its kindof annoying but if I don't add the defaults here and then pass then both as undefined it will override the defaults in `#updateSmartTransaction`.

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts:353]: nope 🤔 also not sure why this change was made...

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts:302]: Yeah perhaps. This was from a while ago but I think my thinking here was optimizing for fewer calls to `this.getNetworkClientById` but it doesn't really matter either way.

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts:616]: good catch! yessir!

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts:751]: yep good catch!

Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/caf5ac0a59b73823ef6977cb480bf4c3ec2b5b22

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/caf5ac0a59b73823ef6977cb480bf4c3ec2b5b22

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/caf5ac0a59b73823ef6977cb480bf4c3ec2b5b22

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/caf5ac0a59b73823ef6977cb480bf4c3ec2b5b22

File: src/SmartTransactionsController.test.ts
Comment (adonesky1) [reply src/SmartTransactionsController.test.ts]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/caf5ac0a59b73823ef6977cb480bf4c3ec2b5b22

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: this can/should be removed but I need to move the `this.initialize()` call above. Without moving this we were always using the DEFAULT_INTERVAL. done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/8e2ee0de68987a45c13375a2770300251bf08e7f

File: src/SmartTransactionsController.ts
Comment (jiexi) [inline src/SmartTransactionsController.ts:74]: probably should be private. nbd deal

File: src/SmartTransactionsController.test.ts
Comment (jiexi) [inline src/SmartTransactionsController.test.ts:439]: yeah... have run into issues like this as well when testing these polling controllers. Could try stubbing the prototype before instantiating the controller. Up to you if you want to address here as we found the test in this state

File: src/SmartTransactionsController.test.ts
Comment (jiexi) [inline src/SmartTransactionsController.test.ts:538]: is this needed? I guess the reason why you're doing this is because you're not confident that outside polling hasn't changed this for us?

File: src/SmartTransactionsController.test.ts
Comment (jiexi) [inline src/SmartTransactionsController.test.ts:903]: would it be to tedious to use matchers and `expect(spy).toHaveBeenNthCalledWith()` instead of `mockClear`?

File: src/SmartTransactionsController.test.ts
Comment (adonesky1) [reply src/SmartTransactionsController.test.ts:538]: yeah def not strictly needed, I just tend to add these extra assertions to make sure that the later assertions are still testing what they think they are.

File: src/SmartTransactionsController.test.ts
Comment (adonesky1) [reply src/SmartTransactionsController.test.ts:903]: sure that works!

Comment (adonesky1) [reply src/SmartTransactionsController.test.ts:903]: done here: https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/2b08ccd27a429deaad6d7e318a613ab5b750626b

File: src/SmartTransactionsController.ts
Comment (adonesky1) [reply src/SmartTransactionsController.ts]: done here https://github.com/MetaMask/smart-transactions-controller/pull/237/commits/2b08ccd27a429deaad6d7e318a613ab5b750626b

**Review by jiexi — APPROVED** [review]
🚀

**Comment by adonesky1** [comment]
Going to wait for a review from @dan437 before merging

**Pull request merged**
